File: web/src/lib/difficulty.ts
import type { DifficultyId } from "../types";

export interface DifficultyConfig {
  id: DifficultyId;
  timerRanked: number | null;
  timerCasual: number | null;
  startZoom: number;
  maxZoom: number;
  scoreMultiplier: number;
  showRegion: boolean;
  showCountryAfter: boolean;
  rounds: number;
  movingPenalty: boolean;
  noLook: boolean;
}

export const DIFFICULTIES: Record<DifficultyId, DifficultyConfig> = {
  easy: {
    id: "easy",
    timerRanked: null,
    timerCasual: null,
    startZoom: 3,
    maxZoom: 12,
    scoreMultiplier: 0.9,
    showRegion: true,
    showCountryAfter: true,
    rounds: 5,
    movingPenalty: false,
    noLook: false,
  },
  medium: {
    id: "medium",
    timerRanked: 180,
    timerCasual: null,
    startZoom: 2,
    maxZoom: 15,
    scoreMultiplier: 1,
    showRegion: false,
    showCountryAfter: true,
    rounds: 5,
    movingPenalty: false,
    noLook: false,
  },
  hard: {
    id: "hard",
    timerRanked: 120,
    timerCasual: null,
    startZoom: 2,
    maxZoom: 11,
    scoreMultiplier: 1.2,
    showRegion: false,
    showCountryAfter: true,
    rounds: 5,
    movingPenalty: true,
    noLook: false,
  },
  impossible: {
    id: "impossible",
    timerRanked: 60,
    timerCasual: 75,
    startZoom: 1,
    maxZoom: 8,
    scoreMultiplier: 1.5,
    showRegion: false,
    showCountryAfter: false,
    rounds: 5,
    movingPenalty: true,
    noLook: true,
  },
};

export const DUEL_ROUNDS = 5;

export const DIFFICULTY_ORDER: DifficultyId[] = ["easy", "medium", "hard", "impossible"];

export function getDifficulty(id: DifficultyId): DifficultyConfig {
  return DIFFICULTIES[id] ?? DIFFICULTIES.medium;
}

File: src/systems/difficulty.py
"""Difficulty presets — GeoGuessr-style game modes."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Difficulty:
    id: str
    timer_ranked: int | None
    timer_casual: int | None
    start_zoom: int
    max_zoom: int
    score_multiplier: float
    show_region: bool
    show_country_after: bool
    rounds: int
    moving_penalty: bool


DIFFICULTIES: dict[str, Difficulty] = {
    "easy": Difficulty(
        id="easy",
        timer_ranked=None,
        timer_casual=None,
        start_zoom=3,
        max_zoom=12,
        score_multiplier=0.9,
        show_region=True,
        show_country_after=True,
        rounds=5,
        moving_penalty=False,
    ),
    "medium": Difficulty(
        id="medium",
        timer_ranked=180,
        timer_casual=None,
        start_zoom=2,
        max_zoom=15,
        score_multiplier=1.0,
        show_region=False,
        show_country_after=True,
        rounds=5,
        moving_penalty=False,
    ),
    "hard": Difficulty(
        id="hard",
        timer_ranked=120,
        timer_casual=None,
        start_zoom=2,
        max_zoom=11,
        score_multiplier=1.2,
        show_region=False,
        show_country_after=True,
        rounds=5,
        moving_penalty=True,
    ),
    "impossible": Difficulty(
        id="impossible",
        timer_ranked=60,
        timer_casual=75,
        start_zoom=1,
        max_zoom=8,
        score_multiplier=1.5,
        show_region=False,
        show_country_after=False,
        rounds=5,
        moving_penalty=True,
    ),
}

ORDER = ("easy", "medium", "hard", "impossible")


def get_difficulty(diff_id: str) -> Difficulty:
    return DIFFICULTIES.get(diff_id, DIFFICULTIES["medium"])
